commit 834bc27940507737f064bca7362b4c7f4cd54c98
Author: John Spray <jspray@redhat.com>
Date:   Thu Nov 23 08:01:33 2017 -0500

    mgr: load command definitions earlier
    
    ...and for all modules, not just the active ones.
    
    This enables us to give better feedback to the user
    when they try and use a command from a disabled module,
    and also fixes the race between enabling a module and
    trying to use its commands.
    
    Fixes: http://tracker.ceph.com/issues/21683
    Signed-off-by: John Spray <john.spray@redhat.com>

diff --git a/src/mgr/ActivePyModules.cc b/src/mgr/ActivePyModules.cc
index 76b54203ded..7b370800926 100644
--- a/src/mgr/ActivePyModules.cc
+++ b/src/mgr/ActivePyModules.cc
@@ -491,34 +491,6 @@ void ActivePyModules::set_config(const std::string &module_name,
   }
 }
 
-std::vector<ModuleCommand> ActivePyModules::get_py_commands() const
-{
-  Mutex::Locker l(lock);
-
-  std::vector<ModuleCommand> result;
-  for (const auto& i : modules) {
-    auto module = i.second.get();
-    auto mod_commands = module->get_commands();
-    for (auto j : mod_commands) {
-      result.push_back(j);
-    }
-  }
-
-  return result;
-}
-
-std::vector<MonCommand> ActivePyModules::get_commands() const
-{
-  std::vector<ModuleCommand> commands = get_py_commands();
-  std::vector<MonCommand> result;
-  for (auto &pyc: commands) {
-    result.push_back({pyc.cmdstring, pyc.helpstring, "mgr",
-                        pyc.perm, "cli", MonCommand::FLAG_MGR});
-  }
-  return result;
-}
-
-
 std::map<std::string, std::string> ActivePyModules::get_services() const
 {
   std::map<std::string, std::string> result;
@@ -713,6 +685,18 @@ void ActivePyModules::set_health_checks(const std::string& module_name,
   }
 }
 
+int ActivePyModules::handle_command(
+  std::string const &module_name,
+  const cmdmap_t &cmdmap,
+  std::stringstream *ds,
+  std::stringstream *ss)
+{
+  lock.Lock();
+  auto mod = modules.at(module_name).get();
+  lock.Unlock();
+  return mod->handle_command(cmdmap, ds, ss);
+}
+
 void ActivePyModules::get_health_checks(health_check_map_t *checks)
 {
   Mutex::Locker l(lock);
diff --git a/src/mgr/ActivePyModules.h b/src/mgr/ActivePyModules.h
index 6d0feedb417..60572c25bfc 100644
--- a/src/mgr/ActivePyModules.h
+++ b/src/mgr/ActivePyModules.h
@@ -90,11 +90,11 @@ public:
 
   void set_uri(const std::string& module_name, const std::string &uri);
 
-  // Python command definitions, including callback
-  std::vector<ModuleCommand> get_py_commands() const;
-
-  // Monitor command definitions, suitable for CLI
-  std::vector<MonCommand> get_commands() const;
+  int handle_command(
+    const std::string &module_name,
+    const cmdmap_t &cmdmap,
+    std::stringstream *ds,
+    std::stringstream *ss);
 
   std::map<std::string, std::string> get_services() const;
 
